lib/scholar/neighbors/k_nearest_neighbors.ex

```diff
@@ -0,0 +1,401 @@
+    }
+  end
+
+  @doc """
+  Makes predictions with the given `model` on inputs `x`.
+
+  ## Return Values
+
+    It returns a tensor with predicted class labels
+
+  ## Examples
+
+      iex> model =
+      iex>  Scholar.Neighbors.KNearestNeighbors.fit(Nx.tensor([[1, 2], [2, 4], [1, 3], [2, 5]]), Nx.tensor([1, 0, 1, 1]),
+      ...>    num_classes: 2
+      ...>  )
+      iex> Scholar.Neighbors.KNearestNeighbors.predict(model, Nx.tensor([[1.9, 4.3], [1.1, 2.0]]))
+      #Nx.Tensor<
+        s64[2]
+        [1, 1]
+      >
+  """
+  defn predict(%__MODULE__{labels: labels, weights: weights, task: task} = model, x) do
+    {neigh_distances, neigh_indices} = k_neighbors(model, x)
+    pred_labels = Nx.take(labels, neigh_indices)
+
+    case task do
+      :classification ->
+        case weights do
+          :distance -> mode_weighted(pred_labels, 1 / neigh_distances, axis: 1)
+          :uniform -> Nx.mode(pred_labels, axis: 1)
+        end
+
+      :regression ->
+        case weights do
+          :distance ->
+            if Nx.rank(labels) == 2,
+              do: Nx.weighted_mean(pred_labels, 1 / neigh_distances, axes: [1, 0]),
+              else: Nx.weighted_mean(pred_labels, 1 / neigh_distances, axes: [1])
+
+          :uniform ->
+            Nx.mean(pred_labels, axes: [1])
+        end
+    end
+  end
+
+  @doc """
+  Return probability estimates for the test data `x`.
+
+  ## Return Values
+
+    It returns a tensor with probabilities of classes. They are arranged in lexicographic order.
+
+  ## Examples
+
+      iex> model =
+      iex>  Scholar.Neighbors.KNearestNeighbors.fit(Nx.tensor([[1, 2], [2, 4], [1, 3], [2, 5]]), Nx.tensor([1, 0, 1, 1]),
+      ...>    num_classes: 2
+      ...>  )
+      iex> Scholar.Neighbors.KNearestNeighbors.predict_proba(model, Nx.tensor([[1.9, 4.3], [1.1, 2.0]]))
+      #Nx.Tensor<
+        f32[2][2]
+        [
+          [0.75, 0.25],
+          [0.75, 0.25]
+        ]
+      >
+  """
+  deftransform predict_proba(
+                 %__MODULE__{
+                   task: :classification
+                 } = model,
+                 x
+               ) do
+    predict_proba_n(model, x)
+  end
+
+  defnp predict_proba_n(
+          %__MODULE__{
+            labels: labels,
+            weights: weights,
+            num_classes: num_classes
+          } = model,
+          x
+        ) do
+    {num_samples, _} = Nx.shape(x)
+    {neigh_distances, neigh_indices} = k_neighbors(model, x)
+    pred_labels = Nx.take(labels, neigh_indices)
+    proba = Nx.broadcast(0.0, {num_samples, num_classes})
+
+    weights_vals =
+      case weights do
+        :distance -> 1 / neigh_distances
+        :uniform -> Nx.broadcast(1.0, Nx.shape(neigh_indices))
+      end
+
+    indices =
+      Nx.stack(
+        [Nx.iota(Nx.shape(pred_labels), axis: 0), Nx.take(labels, pred_labels)],
+        axis: -1
+      )
+      |> Nx.flatten(axes: [0, 1])
+
+    proba = Nx.indexed_add(proba, indices, Nx.flatten(weights_vals))
```

Comment:
```suggestion
    proba = Nx.indexed_put(proba, indices, Nx.flatten(weights_vals))
```

indexed_put seems a bit more semantic given it's all zeros anyway

Reply:
nevermind this comment :)